File: app/src/main/java/com/ternaryop/tumblr/TumblrPhotoPost.java
package com.ternaryop.tumblr;

import java.util.ArrayList;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class TumblrPhotoPost extends TumblrPost {
    /**
     * 
     */
    private static final long serialVersionUID = 8910912231608271421L;
    
    private List<TumblrPhoto> photos;
    private String caption;

    public TumblrPhotoPost() {
        super();
    }

    public TumblrPhotoPost(JSONObject json) throws JSONException {
        super(json);

        caption = json.getString("caption");

        JSONArray jsonPhotos = json.getJSONArray("photos");
        photos = new ArrayList<>(jsonPhotos.length());
        for (int i = 0; i < jsonPhotos.length(); i++) {
            photos.add(new TumblrPhoto(jsonPhotos.getJSONObject(i)));
        }
    }

    public TumblrPhotoPost(TumblrPhotoPost photoPost) {
        super(photoPost);
        photos = photoPost.photos;
        caption = photoPost.caption;
    }

    public List<TumblrPhoto> getPhotos() {
        return photos;
    }

    public void setPhotos(List<TumblrPhoto> photos) {
        this.photos = photos;
    }

    public String getCaption() {
        return caption;
    }

    public void setCaption(String caption) {
        this.caption = caption;
    }

    public List<TumblrAltSize> getFirstPhotoAltSize() {
        return photos.size() > 0 ? photos.get(0).getAltSizes() : null;
    }
    
    public TumblrAltSize getClosestPhotoByWidth(int width) {
        for (TumblrAltSize p : photos.get(0).getAltSizes()) {
            // some images don't have the exact (==) width so we get closest width (<=)
            if (p.getWidth() <= width) {
                return p;
            }
        }
        return null;
    }
    
    @Override
    public String toString() {
        return caption;
    }
}

File: app/src/main/java/com/ternaryop/tumblr/TumblrPhotoPost.kt
package com.ternaryop.tumblr

import org.json.JSONException
import org.json.JSONObject

open class TumblrPhotoPost : TumblrPost {

    private var photos: MutableList<TumblrPhoto> = mutableListOf()
    var caption = ""

    val firstPhotoAltSize: List<TumblrAltSize>?
        get() = if (photos.isEmpty()) null else photos[0].altSizes

    constructor()

    @Throws(JSONException::class)
    constructor(json: JSONObject) : super(json) {
        caption = json.getString("caption")

        val jsonPhotos = json.getJSONArray("photos")
        for (i in 0 until jsonPhotos.length()) {
            photos.add(TumblrPhoto(jsonPhotos.getJSONObject(i)))
        }
    }

    constructor(photoPost: TumblrPhotoPost) : super(photoPost) {
        photos = photoPost.photos
        caption = photoPost.caption
    }

    fun getClosestPhotoByWidth(width: Int): TumblrAltSize? {
        // some images don't have the exact (==) width so we get closest width (<=)
        return firstPhotoAltSize?.firstOrNull { it.width <= width }
    }

    override fun toString() = caption

    companion object {
        /**
         *
         */
        private const val serialVersionUID = 8910912231608271421L
    }
}
